confgenerator/confmerger.go

```diff
@@ -0,0 +1,281 @@
+					},
+				},
+			},
+		},
+	}
+)
+
+func MergeConfFiles(userConfPath, confDebugFolder, platform string) error {
+	builtInConfPath := filepath.Join(confDebugFolder, "built-in-config.yaml")
+	mergedConfPath := filepath.Join(confDebugFolder, "merged-config.yaml")
+	return mergeConfFiles(builtInConfPath, userConfPath, mergedConfPath, platform)
+}
+
+func mergeConfFiles(builtInConfPath, userConfPath, mergedConfPath, platform string) error {
+	builtInStruct := builtInConfStructs[platform]
+	builtInYaml, err := yaml.Marshal(builtInStruct)
+	if err != nil {
+		return fmt.Errorf("failed to convert the built-in config %q to yaml: %w \n", builtInConfPath, err)
+	}
+
+	// Write the built-in conf to disk for debugging purpose.
+
+	if err := WriteConfigFile(builtInYaml, builtInConfPath); err != nil {
+		return err
+	}
+
+	// Read the built-in config file.
+	original, err := builtInStruct.DeepCopy()
+	if err != nil {
+		return err
+	}
+
+	// Optionally merge the user config file.
+	if _, err = os.Stat(userConfPath); err != nil {
+		if os.IsNotExist(err) {
+			// Skip the merge if the user config file does not exist.
+		} else {
+			return fmt.Errorf("failed to retrieve the user config file %q: %w \n", userConfPath, err)
+		}
+	} else {
+		overrides, err := ReadUnifiedConfigFromFile(userConfPath)
+		if err != nil {
+			return err
+		}
+		mergeConfigs(&original, &overrides)
+	}
+
+	// Write the merged conf file.
+	configBytes, err := yaml.Marshal(original)
+	if err != nil {
+		return fmt.Errorf("failed to convert the merged config %q to yaml: %w \n", mergedConfPath, err)
+	}
+	if err := ioutil.WriteFile(mergedConfPath, configBytes, 0644); err != nil {
+		return fmt.Errorf("failed to write the merged config file %q: %w \n", mergedConfPath, err)
+	}
+	return nil
+}
+
+func mergeConfigs(original, overrides *UnifiedConfig) {
+	// For "default_pipeline", we go one level deeper.
+	// this covers 2 cases:
+	// 1. if "<receivers / processors / exporters>: []" is specified explicitly in user config, the entity gets cleared.
+	// 2. if "<receivers / processors / exporters>" is a non-empty list, it overrides the built-in list.
+	//    e.g. users might use the config below to turn off iis metrics on windows.
+	//    metrics:
+	//      service:
+	//        pipelines:
+	//          default_pipeline:
+	//            receivers: [hostmetrics,mssql]
+	if overrides.Logging != nil {
+		// Overrides logging.receivers.
+		for k, v := range overrides.Logging.Receivers {
+			original.Logging.Receivers[k] = v
+		}
+
+		// Overrides logging.processors.
+		original.Logging.Processors = map[string]*LoggingProcessor{}
+		for k, v := range overrides.Logging.Processors {
+			original.Logging.Processors[k] = v
+		}
+
+		// Overrides logging.exporters.
+		for k, v := range overrides.Logging.Exporters {
+			original.Logging.Exporters[k] = v
+		}
+		if overrides.Logging.Service != nil {
+			for name, pipeline := range overrides.Logging.Service.Pipelines {
+				if name == "default_pipeline" {
```

Comment:
Why doesn't merging this at the pipeline level just like the other object types produce exactly the same behavior in fewer lines of code?

Reply:
Because of cases like:

```
$ cat testdata/valid/windows/metrics-turn_off_iis/input.yaml

metrics:
  service:
    pipelines:
      default_pipeline:
        receivers: [hostmetrics,mssql]
```

Reply:
I don't see how that is different. That looks like it would merge the same either way.

Reply:
If it's at the pipeline level, users would have to specify the full default_pipeline (receivers, processors, exporters), right?

Reply:
Sample case: https://github.com/GoogleCloudPlatform/ops-agent/blob/1f661798d2c136c156ec3fd45c429269fc00cfc5/confgenerator/testdata/invalid/linux/metrics-processor_exclude_metrics_invalid_parameter_metric_prefixes/input.yaml

```
metrics:
  processors:
    metrics_filter:
      type: exclude_metrics
      metrics_pattern:
      - agent.googleapis.com/processes/not_end_with_forward_slash_with_star
      - not.agent.googleapis.com/processes/*
  service:
    pipelines:
      default_pipeline:
        processors: [metrics_filter]
```

The alternative is for users to always repeat receivers:

```
metrics:
  processors:
    metrics_filter:
      type: exclude_metrics
      metrics_pattern:
      - agent.googleapis.com/processes/not_end_with_forward_slash_with_star
      - not.agent.googleapis.com/processes/*
  service:
    pipelines:
      default_pipeline:
        receivers: [hostmetrics]
        processors: [metrics_filter]
```